src/operations/execute_operation.ts

```diff
@@ -94,63 +96,56 @@ export function executeOperation<
         owner = Symbol();
         session = topology.startSession({ owner, explicit: false });
       } else if (session.hasEnded) {
-        return cb(new MongoExpiredSessionError('Use of expired sessions is not permitted'));
+        return callback(new MongoExpiredSessionError('Use of expired sessions is not permitted'));
       } else if (session.snapshotEnabled && !topology.capabilities.supportsSnapshotReads) {
-        return cb(new MongoCompatibilityError('Snapshot reads require MongoDB 5.0 or later'));
+        return callback(new MongoCompatibilityError('Snapshot reads require MongoDB 5.0 or later'));
       }
     } else if (session) {
       // If the user passed an explicit session and we are still, after server selection,
       // trying to run against a topology that doesn't support sessions we error out.
-      return cb(new MongoCompatibilityError('Current topology does not support sessions'));
+      return callback(new MongoCompatibilityError('Current topology does not support sessions'));
     }
 
     try {
-      executeWithServerSelection(topology, session, operation, (err, result) => {
-        if (session && session.owner && session.owner === owner) {
-          return session.endSession(err2 => cb(err2 || err, result));
+      executeWithServerSelection<TResult>(topology, session, operation, (error, result) => {
+        if (session && session.owner != null && session.owner === owner) {
+          return session.endSession(endSessionError => callback(endSessionError ?? error, result));
         }
 
-        cb(err, result);
+        callback(error, result);
       });
-    } catch (e) {
-      if (session && session.owner && session.owner === owner) {
+    } catch (error) {
+      // TODO: we shouldn't catch here.
+      // We catch and NOT finally, cus its only in case of an error that we want to end the session
```

Comment:
I'm not following this logic - and why do we no longer want to throw the error or pass it to the callback? Looks like it's meant to catch sync errors from executeWithServerSelection?

Reply:
Good catch, I didn't mean to change the logic here, and yea that's what I'm getting at with my quick TODO, this relates to what we found in NODE-3422. This is just something I think we'll address with async/await, having multiple ways to recieve errors is tough to reason about.